src/ptf/dataplane.py

```diff
@@ -694,6 +703,107 @@ def packets(self, device, port=None):
+            self.expected_packet = expected_packet
+            return self
+
+        def format(self):
+            """
+            Returns a string containing a nice (but verbose) representation of
+            this packet. If the expected packet is a scapy packet, it's used to
+            include detailed information about the fields in the packet.
+            """
+            try:
+                # The scapy packet dissection methods print directly to stdout,
+                # so we have to redirect stdout to a string.
+                sys.stdout = StringIO()
+
+                print "========== RECEIVED =========="
+                if isinstance(self.expected_packet, scapy.packet.Packet):
+                    # Dissect this packet as if it were an instance of
+                    # the expected packet's class.
+                    scapy.packet.ls(self.expected_packet.__class__(self.packet))
+                    print '--'
+                scapy.utils.hexdump(self.packet)
+                print "=============================="
+
+                return sys.stdout.getvalue()
+            finally:
+                sys.stdout.close()
+                sys.stdout = sys.__stdout__  # Restore the original stdout.
+
+    class PollFailure(PollResult):
+        """
+        Returned by poll() when it fails to match any packets. Contains metadata
+        which can be used to diagnose the failure; callers will often want to
+        include the result of format() in assertion failure messages.
+
+        For backwards compatibility, when a PollFailure is treated as a tuple
+        and unpacked, all PollResult fields are 'None'.
+        """
+        def __new__(cls, expected_packet, recent_packets, packet_count):
+            self = super(DataPlane.PollFailure, cls).__new__(cls, None, None, None, None)
+            self.expected_packet = expected_packet
+            self.recent_packets = recent_packets
+            self.packet_count = packet_count
+            return self
+
+        def format(self):
+            """
+            Returns a string containing a nice (but verbose) error report based
+            on this PollFailure. If there was an expected packet, it's included
+            in the output. If the expected packet is a scapy packet object, the
+            output will include information about the fields in the packet.
+            """
+            try:
+                # The scapy packet dissection methods print directly to stdout,
+                # so we have to redirect stdout to a string.
+                sys.stdout = StringIO()
+
+                if self.expected_packet is not None:
+                    print "========== EXPECTED =========="
+                    if isinstance(self.expected_packet, scapy.packet.Packet):
+                        scapy.packet.ls(self.expected_packet)
+                        print '--'
+                    scapy.utils.hexdump(self.expected_packet)
```

Comment:
does it still work / make sense if `expected_packet` is an instance of `mask.Mask`